### DISCUSSION TITLE
What is the right way to call an APIRouter endpoint from another in same app

### DISCUSSION THREAD

**Discussion by jintolonappan** [fastapi/fastapi#7006] [Questions] [answered]
### First check

* [x] I added a very descriptive title to this issue.
* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [x] I already read and followed all the tutorial in the docs and didn't find an answer.
* [x] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [x] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [x] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [x] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

## Question
I'm working on creating an application by following the tutorial. I currently have a structure similar 
to below as mentioned in [Bigger Applications - Multiple Files](https://fastapi.tiangolo.com/tutorial/bigger-applications/?h=apirout#include-the-apirouters-for-users-and-items) section of the tutorial:

```
.
├── app
│   ├── __init__.py
│   ├── main.py
│   ├── dependencies.py
│   └── routers
│   │   ├── __init__.py
│   │   ├── items.py
│   │   └── users.py
│   └── internal
│       ├── __init__.py
│       └── admin.py
```

I have a usecase where I need to hit one of the endpoint exposed by one APIRouter (Say, in this example `users` from the other APIRouter (`items` from the example).  

This is required due to some of the foreign key relationships and wanted to POST the data to referenced table rather than throwing an exception back to the user.   
Using above example, the end consumer may send both user data and items data in a single API call and internally, my app needs to hit both `users` endpoint and `items` endpoint.  

I already looked at the [Requests](https://fastapi.tiangolo.com/advanced/using-request-directly/?h=client#use-the-request-object-directly) section and thought there must be a better way rather than figuring out client host and manually coming up with the API endpoint.   
I also looked at [Callback](https://fastapi.tiangolo.com/advanced/openapi-callbacks/) section. 

I'm using `SQLAlchemy` for database connections and is referenced in each APIRouters as `Depends`. 

### Environment
* OS: macOS Catalina 10.15.7
* FastAPI Version: 0.63.0
* Python version: Python 3.8.5

**Comment by insomnes** [comment]
@jintolonappan hello!
Why do you want to call it by HTTP API endpoint request and not by calling some function in app code? 
Wrap the desired actions in function somewhere and call it from endpoint handlers.

**Comment by ycd** [comment, marked as answer]
You can abstract your endpoints logic. Then when you need to call it internally, for example just use the `get_user_from_database_by_id` function.

```python

async def get_user_from_database_by_id(id: int) -> User:
    ...

@app.get("/user/{id}")
async def get_user_by_id(id: int):
    return await get_user_from_database_by_id(id)

**Comment by jintolonappan** [comment]
Thanks @insomnes @ycd
I implemented the call with function calls. The reason I was a little hesitant to use this option was because I thought I could keep each router on its own without cross-router calls.

**Comment by tiangolo** [comment]
Thanks for the help here @ycd and @insomnes ! 👏  🙇 

Thanks for reporting back and closing the issue @jintolonappan  👍

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.
